### PR TITLE
Make k-means example's custom kernels simpler

### LINKED ISSUE #2143: Use `ReductionKernel` in k-means example
The k-mean example uses `ElementwiseKernel` as an example of a custom kernel, but its algorithm is essentially reduction rather than mapping. It would be better to use `ReductionKernel` in the k-means example and to provide another example that uses `ElementwiseKernel`.

### PR DESCRIPTION
**Pull request by takagi**
Close #2143.

This PR makes the k-means example's custom kernels simpler. Additionally, it makes the algorithm computing the centroids completely parallel.

### CODE REVIEW

File: examples/kmeans/kmeans.py [hunk 1]
```diff
@@ -49,35 +45,32 @@ def fit(X, n_clusters, max_iter, use_custom_kernel):
...
+                                   centers[None, :, 1], centers[None, :, 0])
 
         new_pred = xp.argmin(distances, axis=1).astype(np.int32)
         if xp.all(new_pred == pred):
             break
         pred = new_pred
 
         # calculate centers
+        i = xp.arange(n_clusters)
+        mask = pred == i[:, None]
```
[the remark this answers is not in the document]
Comment (takagi) [reply examples/kmeans/kmeans.py]: Can I answer your comment at 6072c26 ?

[the remark this answers is not in the document]
Comment (asi1024) [reply examples/kmeans/kmeans.py]: Hmm, this version seems slower than the original one. I'm sorry but could you revert this change?

[the remark this answers is not in the document]
Comment (takagi) [reply examples/kmeans/kmeans.py]: Thanks for checking the performance. I reverted the change!

**Comment by asi1024** [comment]
Jenkins, test this please.

**Comment by asi1024** [comment]
LGTM!

**Pull request merged**